pkg/templates/pdb/template.go

```diff
@@ -0,0 +1,97 @@
+	"fmt"
+
+	"golang.stackrox.io/kube-linter/pkg/check"
+	"golang.stackrox.io/kube-linter/pkg/config"
+	"golang.stackrox.io/kube-linter/pkg/diagnostic"
+	"golang.stackrox.io/kube-linter/pkg/lintcontext"
+	"golang.stackrox.io/kube-linter/pkg/objectkinds"
+	"golang.stackrox.io/kube-linter/pkg/templates"
+	"golang.stackrox.io/kube-linter/pkg/templates/pdb/internal/params"
+	v1 "k8s.io/api/apps/v1"
+	pdbV1 "k8s.io/api/policy/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	templateKey           = "pod-disruption-budget"
+	maxUnavailableZeroMsg = "MaxUnavailable is set to 0"
+)
+
+func init() {
+	templates.Register(check.Template{
+		HumanName:   "Pod Disruption Budget Configuration Checks",
+		Key:         templateKey,
+		Description: "Flag non-optimal PDB configurations",
+		SupportedObjectKinds: config.ObjectKindsDesc{
+			ObjectKinds: []string{
+				objectkinds.PodDisruptionBudget},
+		},
+		Parameters:             params.ParamDescs,
+		ParseAndValidateParams: params.ParseAndValidate,
+		Instantiate: params.WrapInstantiateFunc(func(p params.Params) (check.Func, error) {
+			return func(lintCtx lintcontext.LintContext, object lintcontext.Object) []diagnostic.Diagnostic {
+
+				var results []diagnostic.Diagnostic
+
+				// Get the PDB provided
+				pdb, ok := object.K8sObject.(*pdbV1.PodDisruptionBudget)
+				if !ok {
+					return nil
+				}
+
+				// If MaxUnavailable is set to 0,
+				// return as no other checks will uniquely apply
+				if pdb.Spec.MaxUnavailable.IntVal == 0 {
+					return []diagnostic.Diagnostic{{
+						Message: maxUnavailableZeroMsg,
+					},
+					}
+				}
+
+				labelSelector, err := metaV1.LabelSelectorAsSelector(&metaV1.LabelSelector{
+					MatchLabels:      pdb.Spec.Selector.MatchLabels,
+					MatchExpressions: pdb.Spec.Selector.MatchExpressions})
+
+				if err != nil {
+					return []diagnostic.Diagnostic{{
+						Message: fmt.Sprintf("pdb has invalid podSelector: %v", err),
+					}}
+				}
+
+				// Evaluate deployments in the lintContext to see if they have MinAvailable set too low
+				for _, deployment := range getDeployments(lintCtx, pdb.Namespace, labelSelector) {
+					if *deployment.Spec.Replicas == pdb.Spec.MinAvailable.IntVal {
+						results = append(results, diagnostic.Diagnostic{
+							Message: fmt.Sprintf("Deployment %s has replicas equal to the minumum available replicas set by its PDB.", deployment.Name),
+						})
+					}
+				}
+				return results
+
+			}, nil
+		}),
+	})
+}
+
+func getDeployments(lintCtx lintcontext.LintContext, namespace string, labelSelector labels.Selector) []v1.Deployment {
+
+	var deploymentList []v1.Deployment
+
+	for _, obj := range lintCtx.Objects() {
+		deployment, ok := obj.K8sObject.(*v1.Deployment)
+		if !ok {
+			continue
+		}
+
+		// Find any deployments in the same namespace with labels that match the PDBs selector
+		if labelSelector.Matches(labels.Set(deployment.Labels)) && namespace == deployment.Namespace {
```

Comment:
Just question @fahlmant please. This requires the deployment to have the labels set, right? Wouldn't be better to check `labelSelector.Matches(labels.Set(deployment.Spec.Selector.MatchLabels))` ?

Reply:
You're right, I was thinking about the pods the PDB checks, rather than the deployment backing the pods. Fixing